docs/articles/documentation/extending-testcafe/custom-browser-provider-plugin/index.md

```diff
@@ -0,0 +1,180 @@
+The generator will scaffold out the plugin, so that you only need to write a few lines of code.
+
+For more information on how to implement the plugin, see the following sections.
+
+* [Generating Browser Provider Project](#generating-browser-provider-project)
+* [Implementing the Browser Provider](#implementing-the-browser-provider)
+* [Building the Provider](#building-the-provider)
+* [Using the Provider Development Version](#using-the-provider-development-version)
+* [Publishing the Provider to NPM](#publishing-the-provider-to-npm)
+* [Reference](#reference)
+
+## Generating Browser Provider Project
+
+First, install [Yeoman](http://yeoman.io) and `generator-testcafe-browser-provider` using [npm](https://www.npmjs.com/).
+
+```bash
+npm install -g yo
+npm install -g generator-testcafe-browser-provider
+```
+
+Create a new directory where the generator should place your scaffolded project files and go into it.
+
+```bash
+mkdir my-browser-provider
+cd my-browser-provider
+```
+
+> It is recommended to name the directory as you would like to name the browser provider project. When you run the generator, it will automatically suggest the browser provider name that matches the browser provider directory name. As for the provider package name, it will consist of two parts - the `{testcafe-browser-provider-}` prefix and the name of the provider itself, for example, `{testcafe-browser-provider-my-provider}`.
+
+Then run the browser provider generator to create a new project:
+
+```bash
+yo testcafe-browser-provider
+```
+
+The generator will provide additional settings to customize your browser provider.
+Then Yeoman will automatically scaffold out your browser provider, install the required dependencies, and pull in several useful Gulp tasks for your workflow.
+
+## Implementing the Browser Provider
+
+Once the browser provider has been scaffolded out, go to the provider directory and open the `src/index.js` file.
+
+There are several required and optional provider methods you can implement.
+
+```js
+// Required methods
+async openBrowser (/* id, browserInfo, startPage */) {
+    throw new Error('Not implemented!');
+},
+
+async closeBrowser (/* id, pageInfo */) {
+    throw new Error('Not implemented!');
+},
+
+// Optional initialization and cleanup methods
+async init () {
+    return;
+},
+
+async dispose () {
+    return;
+},
+
+// Optional methods for multi-browser support
+async getBrowserList () {
+    throw new Error('Not implemented!');
+},
+
+async isValidBrowserName (/* browserName */) {
+    return true;
+},
+
+
+// Optional extra methods
+async resizeWindow (/* id, pageInfo, width, height */) {
+    this.reportWarning('The screenshot functionality is not supported by the "<%= providerName %>" browser provider.');
+},
+
+async takeScreenshot (/* id, pageInfo, screenshotPath */) {
+    this.reportWarning('The window resize functionality is not supported by the "<%= providerName %>" browser provider.');
```

Comment:
It's not the point of my question. You've copied generator's template. Once generator made it's job `"<%= providerName %>"` should be replaced with the actual provider name. Since, it's an example of generated provider we shouldn't have generator's templates here.